### PR TITLE
UL4.1, UL4.2, UL4.3, UL4.4

### PR DESCRIPTION
**Pull request by krystian-hebel**

### CODE REVIEW

**Comment by miczyg1** [comment]
@krystian-hebel i guess wrong target branch. 17k lines of code

**Comment by krystian-hebel** [comment]
@miczyg1 that is correct. Most of this is copy of ShellPkg.

**Comment by miczyg1** [comment]
@krystian-hebel we should distinguish features/branches. Shell and trampoline stuff maybe should land in boot_to_shell? arbitrary_binary is intended to store code showing integration of arbitrary binaries in SPI flash. If You think that arbitrary_binary is lacking some parts of this change please do the rebase

**Comment by krystian-hebel** [comment]
@miczyg1 trampoline requires that binary is present in image.

**Comment by krystian-hebel** [comment]
@miczyg1 should be fixed now. I wasn't expecting reserved sections in custom_shell.

File: Vlv2TbltDevicePkg/PlatformPkgGcc.fdf [hunk 1]
```diff
@@ -160,10 +160,11 @@ FV = FVMAIN_COMPACT
   #
 $(FLASH_REGION_RESERVED_OFFSET)|$(FLASH_REGION_RESERVED_SIZE)
 gPlatformModuleTokenSpaceGuid.PcdFlashRsvdBase|gPlatformModuleTokenSpaceGuid.PcdFlashRsvdSize
-DATA = {
-  0x52, 0x45, 0x53, 0x45, 0x52, 0x56, 0x45, 0x44,
-  0x00
-}  # "RESERVED" in HEX
+FILE = hello_world.efi
+#DATA = {
```
Comment (miczyg1) [inline Vlv2TbltDevicePkg/PlatformPkgGcc.fdf]: Since this branch is intended to launch target binaries via trampoline, these DATA sections are not necessary

File: Vlv2TbltDevicePkg/PlatformPkgGcc.fdf [hunk 2]
```diff
@@ -160,10 +160,11 @@ FV = FVMAIN_COMPACT
   #
 $(FLASH_REGION_RESERVED_OFFSET)|$(FLASH_REGION_RESERVED_SIZE)
 gPlatformModuleTokenSpaceGuid.PcdFlashRsvdBase|gPlatformModuleTokenSpaceGuid.PcdFlashRsvdSize
-DATA = {
-  0x52, 0x45, 0x53, 0x45, 0x52, 0x56, 0x45, 0x44,
-  0x00
-}  # "RESERVED" in HEX
+FILE = hello_world.efi
```
Comment (miczyg1) [inline Vlv2TbltDevicePkg/PlatformPkgGcc.fdf]: Let's name it `target_binary.efi` maybe

**Review by miczyg1 — CHANGES_REQUESTED** [review]

**Comment by pietrushnic** [comment]
@krystian-hebel modified files requires correct copyrights.

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 3]
```diff
@@ -314,6 +315,219 @@ InternalEfiShellStartCtrlSMonitor(
   return (Status);
 }
 
+/**
+  internal worker function to load and run an image via memory-mapped device path.
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]: @krystian-hebel please be consistent: either you use fits bit letter either not.

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 4]
```diff
@@ -314,6 +315,219 @@ InternalEfiShellStartCtrlSMonitor(
...
+                                environment variable name and y is the value. If this
+                                is NULL, then the current shell environment is used.
+
+  @param[out] StartImageStatus  Returned status from gBS->StartImage.
+
+  @retval EFI_SUCCESS       The command executed successfully. The status code
+                            returned by the command is pointed to by StatusCode.
+  @retval EFI_INVALID_PARAMETER The parameters are invalid.
+  @retval EFI_OUT_OF_RESOURCES Out of resources.
+  @retval EFI_UNSUPPORTED   Nested shell invocations are not allowed.
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]: @krystian-hebel maybe correct alignment

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 5]
```diff
@@ -314,6 +315,219 @@ InternalEfiShellStartCtrlSMonitor(
...
+  UINTN                         Index;
+  CHAR16                        *Walker;
+  CHAR16                        *NewCmdLine;
+
+  if (ParentImageHandle == NULL) {
+    return (EFI_INVALID_PARAMETER);
+  }
+
+  InitializeListHead(&OrigEnvs);
+  ZeroMem(&ShellParamsProtocol, sizeof(EFI_SHELL_PARAMETERS_PROTOCOL));
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:382]: @krystian-hebel there is no returned values from above functions?

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 6]
```diff
@@ -314,6 +315,219 @@ InternalEfiShellStartCtrlSMonitor(
...
+  }
+
+  //
+  // Load the image with:
+  // FALSE - not from boot manager,
+  // address and size of memory region where the image is located
+  //
+  Status = gBS->LoadImage(
+    FALSE,
+    *ParentImageHandle,
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:398]: @krystian-hebel there is no sanity checking for external parameters?

**Review by pietrushnic — CHANGES_REQUESTED** [review]
Didn't finished review yet, but wanted to provide at least this feedback.

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c (hunk 6 above)
Comment (krystian-hebel) [reply Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:398]: LoadImage is foolproof and does enough checks.

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c (hunk 5 above)
Comment (krystian-hebel) [reply Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:382]: No, these are `VOID`s

**Comment by krystian-hebel** [comment]
I've rebased this branch (again) to get rid of unnecessary commits with reserved sections of SPI.

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 7]
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
...
+  EFI_LOADED_IMAGE_PROTOCOL     *LoadedImage;
+  LIST_ENTRY                    OrigEnvs;
+  EFI_SHELL_PARAMETERS_PROTOCOL ShellParamsProtocol;
+  CHAR16                        *ImagePath;
+  UINTN                         Index;
+  CHAR16                        *Walker;
+  CHAR16                        *NewCmdLine;
+
+  if (ParentImageHandle == NULL) {
+    return (EFI_INVALID_PARAMETER);
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]: Why we are not consistent?

```suggestion
    return EFI_INVALID_PARAMETER;
```

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 8]
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
...
+        -1,
+        NULL,
+        STRING_TOKEN (STR_SHELL_IMAGE_NOT_APP),
+        ShellInfoObject.HiiHandle
+      );
+      goto UnloadImage;
+    }
+
+    ASSERT(LoadedImage->LoadOptionsSize == 0);
+    if (NewCmdLine != NULL) {
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]: @krystian-hebel this check seems to be redundant? You already check if NewCmdLine is NULL right after getting this memory. I don't see why you do it it again.

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 9]
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
...
+
+  NewHandle = NULL;
+
+  NewCmdLine = AllocateCopyPool (StrSize (CommandLine), CommandLine);
+  if (NewCmdLine == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  for (Walker = NewCmdLine; Walker != NULL && *Walker != CHAR_NULL ; Walker++) {
+    if (*Walker == L'^' && *(Walker+1) == L'#') {
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]: @krystian-hebel why we look for `^#` ? Can you add comment here?

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 10]
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
...
+        }
+        ShellParamsProtocol.Argc = 1;
+      } else {
+        // Free the string UpdateArgcArgv put in Argv[0];
+        FreePool (ShellParamsProtocol.Argv[0]);
+      }
+      ShellParamsProtocol.Argv[0] = ImagePath;
+    }
+
+    Status = gBS->InstallProtocolInterface(&NewHandle, &gEfiShellParametersProtocolGuid, EFI_NATIVE_INTERFACE, &ShellParamsProtocol);
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]: @krystian-hebel this invocation has inconsistent style in comparison to other calls.

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 11]
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
...
+    ASSERT_EFI_ERROR(Status);
+
+    ///@todo initialize and install ShellInterface protocol on the new image for compatibility if - PcdGetBool(PcdShellSupportOldProtocols)
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]: @krystian-hebel when we plan to implement that? If trampoline is fully functional without it I would recommend to put that into future improvements section.

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 12]
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
...
+
+    //
+    // now start the image and if the caller wanted the return code pass it to them...
+    //
+    if (!EFI_ERROR(Status)) {
+      StartStatus      = gBS->StartImage(
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]:
```suggestion
      StartStatus = gBS->StartImage(
```

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 13]
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
...
+      }
+
+      CleanupStatus = gBS->UninstallProtocolInterface(
+                            NewHandle,
+                            &gEfiShellParametersProtocolGuid,
+                            &ShellParamsProtocol
+                            );
+      ASSERT_EFI_ERROR(CleanupStatus);
+
+      goto FreeAlloc;
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:502]: @krystian-hebel I have concerns that those `gotos` make code convoluted and quite hard to follow.

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 14]
```diff
@@ -314,6 +316,220 @@ InternalEfiShellStartCtrlSMonitor(
...
+
+FreeAlloc:
+    // Free Argv (Allocated in UpdateArgcArgv)
+    if (ShellParamsProtocol.Argv != NULL) {
+      for (Index = 0; Index < ShellParamsProtocol.Argc; Index++) {
+        if (ShellParamsProtocol.Argv[Index] != NULL) {
+          FreePool (ShellParamsProtocol.Argv[Index]);
+        }
+      }
+      FreePool (ShellParamsProtocol.Argv);
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]: @krystian-hebel convention for function calls is not consistent sometimes you have `foo (bar)` and sometimes `foo(bar)`. Coding style use first method.

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c [hunk 15]
```diff
@@ -339,6 +555,22 @@ UefiMain (
...
   EFI_SIMPLE_TEXT_INPUT_PROTOCOL  *OldConIn;
   SPLIT_LIST                      *Split;
+  EFI_STATUS                      StartImageStatus;
+  MEMMAP_DEVICE_PATH              MemPath[2];
+  CHAR16                          *CommandLine = L"RTNicPgX32.efi /efuse /r";
+
+  MemPath[0].Header.Type = HARDWARE_DEVICE_PATH;
+  MemPath[0].Header.SubType = HW_MEMMAP_DP;
+  MemPath[0].Header.Length[0] = (UINT8)sizeof(MEMMAP_DEVICE_PATH);
+  MemPath[0].Header.Length[1] = (UINT8)(sizeof(MEMMAP_DEVICE_PATH)>> 8);
```
Comment (pietrushnic) [inline Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]:
```suggestion
  MemPath[0].Header.Length[1] = (UINT8)(sizeof (MEMMAP_DEVICE_PATH) >> 8);
```

**Review by pietrushnic — CHANGES_REQUESTED** [review]

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c (hunk 12 above)
Comment (pietrushnic) [reply Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]: @krystian-hebel ?

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c (hunk 13 above)
Comment (pietrushnic) [reply Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:502]: @krystian-hebel do we plan any refactor?

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c (hunk 12 above)
Comment (krystian-hebel) [reply Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c]: Ups, missed it.

File: Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c (hunk 13 above)
Comment (krystian-hebel) [reply Vlv2TbltDevicePkg/CustomShellPkg/Application/Shell/Shell.c:502]: I'm not planning to change this. In my opinion repeating cleanup operations after every point of possible error in this code would make unnecessarily deep `if` branches.

**Review by miczyg1 — APPROVED** [review]
